pkg/cli/cachedaemon.go

```diff
@@ -0,0 +1,176 @@
+package cli
+
+import (
+	"context"
+	"crypto/tls"
+	"flag"
+	"fmt"
+	"net"
+	"os"
+	"os/signal"
+	"runtime/pprof"
+	"syscall"
+
+	"github.com/gadget-inc/dateilager/internal/environment"
+	"github.com/gadget-inc/dateilager/internal/key"
+	"github.com/gadget-inc/dateilager/internal/logger"
+	"github.com/gadget-inc/dateilager/internal/telemetry"
+	"github.com/gadget-inc/dateilager/pkg/cached"
+	"github.com/gadget-inc/dateilager/pkg/client"
+	"github.com/gadget-inc/dateilager/pkg/version"
+	"github.com/spf13/cobra"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+func NewCacheDaemonCommand() *cobra.Command {
+	var (
+		profilerEnabled   bool = false
+		shutdownTelemetry func()
+	)
+
+	var (
+		level        *zapcore.Level
+		encoding     string
+		tracing      bool
+		profilePath  string
+		upstreamHost string
+		upstreamPort uint16
+		certFile     string
+		keyFile      string
+		pasetoFile   string
+		port         int
+		timeout      uint
+		headlessHost string
+		stagingPath  string
+	)
+
+	cmd := &cobra.Command{
+		Use:               "cached",
+		Short:             "DateiLager cache daemon",
+		DisableAutoGenTag: true,
+		Version:           version.Version,
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			cmd.SilenceUsage = true // silence usage when an error occurs after flags have been parsed
+
+			env, err := environment.LoadEnvironment()
+			if err != nil {
+				return fmt.Errorf("could not load environment: %w", err)
+			}
+
+			var config zap.Config
+			if env == environment.Prod {
+				config = zap.NewProductionConfig()
+			} else {
+				config = zap.NewDevelopmentConfig()
+			}
+
+			config.Encoding = encoding
+			config.Level = zap.NewAtomicLevelAt(*level)
+			config.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
+
+			err = logger.Init(config)
+			if err != nil {
+				return fmt.Errorf("could not initialize logger: %w", err)
+			}
+
+			ctx := cmd.Context()
+
+			if profilePath != "" {
+				file, err := os.Create(profilePath)
+				if err != nil {
+					return fmt.Errorf("cannot open profile path %s: %w", profilePath, err)
+				}
+				_ = pprof.StartCPUProfile(file)
+				profilerEnabled = true
+			}
+
+			if tracing {
+				shutdownTelemetry = telemetry.Init(ctx, telemetry.Server)
+			}
+
+			cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+			if err != nil {
+				return fmt.Errorf("cannot open TLS cert and key files (%s, %s): %w", certFile, keyFile, err)
+			}
+
+			pasetoKey, err := parsePublicKey(pasetoFile)
+			if err != nil {
+				return fmt.Errorf("cannot parse Paseto public key %s: %w", pasetoFile, err)
+			}
+
+			cl, err := client.NewClient(ctx, upstreamHost, upstreamPort, client.WithheadlessHost(headlessHost))
+			if err != nil {
+				return err
+			}
+			ctx = client.IntoContext(ctx, cl)
```

Comment:
Probably not necessary in this context since you don't have child commands that need to pull out the client, you are using it directly a few lines below.

Reply:
👍